File: backjoon_review/dfs&bfs/7576.py
import sys
from collections import deque

input = sys.stdin.readline


M, N = map(int,input().split())

graph = [ [0] * M for _ in range(N)]

q = deque()

dx = [-1, 1, 0 ,0]
dy = [0, 0, -1, 1]

for y in range(N):
    numList = list(map(int,input().split()))
    for x in range(M):
        graph[y][x] = numList[x]
        if numList[x] == 1:
            q.append((x,y))

while q:
    nodeX, nodeY = q.popleft()

    for i in range(4):
        nx, ny = nodeX + dx[i],  nodeY + dy[i]
        if 0 <= nx < M and 0 <= ny < N and graph[ny][nx] == 0:
            q.append((nx, ny))
            graph[ny][nx] = graph[nodeY][nodeX] + 1

maxValue = -1
isPossible = True
for y in range(N):
    # print(graph[y][:])
    for x in range(M):
        maxValue = max(maxValue, graph[y][x])
        if graph[y][x] == 0:
            isPossible = False

if isPossible:
    print(maxValue - 1)
else:
    print(-1)

File: baekjoon_js/dfs&bfs/7576.js
("use strict");

const fs = require("fs");
const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `2 2
1 -1
-1 1
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

// console.log(cnt);
const [M, N] = input().split(" ").map(Number);

let graph = Array(N)
  .fill(Number(0))
  .map(() => Array(M).fill(Number(0)));

let q = [];

let dx = [1, -1, 0, 0];
let dy = [0, 0, 1, -1];

for (let y = 0; y < N; y++) {
  numList = input().split(" ").map(Number);
  for (let x = 0; x < M; x++) {
    graph[y][x] = numList[x];
    // console.log(numList[x]);
    if (numList[x] === 1) {
      q.push([x, y]);
    }
  }
}

while (q.length > 0) {
  let [x, y] = q.shift();

  for (let i = 0; i < 4; i++) {
    let [nx, ny] = [x + dx[i], y + dy[i]];

    if (nx >= 0 && nx < M && ny >= 0 && ny < N) {
      // console.log(nx, ny, graph[ny][nx], isVisited[ny][nx]);
      if (graph[ny][nx] === 0) {
        graph[ny][nx] = graph[y][x] + 1;
        q.push([nx, ny]);
      }
    }
  }
  // console.log(graph);
}

let maxValue = 0;
let isPossible = true;
graph.forEach((element) => {
  element.forEach((e) => {
    maxValue = Math.max(maxValue, e);
    if (e === 0) isPossible = false;
  });
});

if (isPossible) console.log(maxValue - 1);
else console.log(-1);
